File: backend/src/routes/auth/oauth.rs
//! 认证路由 - OAuth 登录实现

use mongodb::Database;
use rocket::http::{Cookie, CookieJar, SameSite};
use rocket::serde::json::Json;
use rocket::{State, response::Redirect};

use crate::config::OAuthConfig;
use crate::models::ApiResponse;
use crate::repositories::OptionsRepository;
use crate::services::auth::identity::{IdentityService, OAuthUserPayload};
use crate::services::{OAuthProviderType, OAuthService, OAuthUserInfo};

/// OAuth 重定向端点
///
/// 路由: GET /api/auth/oauth/<provider>
#[get("/oauth/<provider>")]
pub async fn oauth_redirect(
    provider: &str,
    config: &State<OAuthConfig>,
    db: &State<Database>,
) -> Result<Redirect, Json<ApiResponse<()>>> {
    log::info!("OAuth 重定向请求: provider={provider}");

    // 1. 获取最新的 OAuth 配置（数据库优先）
    let options_repo = OptionsRepository::new(db);
    let db_oauth_options = options_repo.get_oauth_config().await.map_err(|e| {
        log::error!("读取数据库配置失败: {e}");
        ApiResponse::internal_error("读取配置失败".to_string())
    })?;

    let redirect_url = match provider {
        "github" => {
            let client_id = db_oauth_options
                .github_client_id
                .filter(|s| !s.is_empty())
                .unwrap_or_else(|| config.github_client_id.clone());

            if client_id.is_empty() {
                return Err(ApiResponse::internal_error(
                    "GitHub OAuth 未配置".to_string(),
                ));
            }

            format!(
                "https://github.com/login/oauth/authorize?client_id={}&redirect_uri={}&scope=user:email",
                client_id,
                urlencoding::encode(&config.github_redirect_uri())
            )
        }
        "qq" => {
            // QQ OAuth 使用统一的服务
            let oauth_service = OAuthService::new(None, None, Some(config.qq_redirect_uri()));
            oauth_service.get_qq_authorize_url().unwrap_or_else(|_| {
                format!(
                    "https://api-space.tnxg.top/oauth/qq/authorize?redirect=true&return_url={}",
                    urlencoding::encode(&config.qq_redirect_uri())
                )
            })
        }
        _ => {
            return Err(ApiResponse::bad_request(format!(
                "不支持的提供商: {provider}"
            )));
        }
    };

    Ok(Redirect::to(redirect_url))
}

/// OAuth 回调端点
///
/// 路由: GET /api/auth/oauth/<provider>/callback?code=xxx
#[get("/oauth/<provider>/callback?<code>")]
pub async fn oauth_callback(
    provider: &str,
    code: String,
    config: &State<OAuthConfig>,
    db: &State<Database>,
    cookies: &CookieJar<'_>,
) -> Result<Redirect, Json<ApiResponse<()>>> {
    log::info!("OAuth 回调处理开始: provider={provider}");

    // 1. 获取第三方用户信息并转换为标准 Payload
    let payload_result = match provider {
        "github" => handle_github_logic(&code, config, db).await,
        "qq" => handle_qq_logic(&code, config, db).await,
        _ => Err(ApiResponse::bad_request("不支持的提供商".to_string())),
    };

    // 如果获取第三方信息失败，重定向到前端并带上错误参数
    let payload = match payload_result {
        Ok(p) => p,
        Err(e) => {
            let err_msg = e.0.message;
            return Ok(Redirect::to(format!(
                "{}/auth/callback?error={}",
                config.frontend_url,
                urlencoding::encode(&err_msg)
            )));
        }
    };

    // 2. 使用 IdentityService 处理复杂的业务逻辑 (自动关联、临时账户等)
    let id_service = IdentityService::new(db, config.jwt_secret.clone());
    let (user_id, is_owner, is_new_user) = match id_service.process_oauth_login(payload).await {
        Ok(res) => res,
        Err(e) => {
            return Ok(Redirect::to(format!(
                "{}/auth/callback?error={}",
                config.frontend_url,
                urlencoding::encode(&e)
            )));
        }
    };

    // 3. 颁发 JWT 令牌
    let token = id_service
        .issue_token(user_id, is_owner)
        .map_err(ApiResponse::internal_error)?;

    // 4. 设置 HttpOnly Cookie（用于后端 API 鉴权）
    let mut cookie = Cookie::new("auth_token", token.clone());
    cookie.set_http_only(true);
    cookie.set_secure(true); // 建议生产环境强制 HTTPS
    cookie.set_same_site(SameSite::Lax);
    cookie.set_path("/");
    cookie.set_max_age(rocket::time::Duration::days(7));
    cookies.add(cookie);

    // 5. 重定向回前端页面
    // 前端会解析 URL 中的 token 和 new_user 标记来决定是进入仪表盘还是进入“绑定/跳过”页面
    let callback_url = format!(
        "{}/auth/callback?token={}&new_user={}",
        config.frontend_url, token, is_new_user
    );

    Ok(Redirect::to(callback_url))
}

// --- 内部逻辑封装：负责与各平台 API 交互 ---

/// 将 `OAuthUserInfo` 转换为 `OAuthUserPayload`
fn convert_to_payload(info: OAuthUserInfo) -> OAuthUserPayload {
    let provider = match info.provider {
        OAuthProviderType::GitHub => "github",
        OAuthProviderType::QQ => "qq",
    }
    .to_string();

    OAuthUserPayload {
        provider,
        provider_id: info.provider_user_id,
        name: info.nickname,
        email: info.email,
        avatar: Some(info.avatar),
        handle: None,
        access_token: info.access_token.unwrap_or_default(),
        scope: None,
    }
}

/// 处理 GitHub 的 OAuth 交换逻辑
async fn handle_github_logic(
    code: &str,
    config: &OAuthConfig,
    db: &Database,
) -> Result<OAuthUserPayload, Json<ApiResponse<()>>> {
    let options_repo = OptionsRepository::new(db);
    let db_oauth = options_repo.get_oauth_config().await.unwrap_or_default();

    let client_id = db_oauth
        .github_client_id
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| config.github_client_id.clone());
    let client_secret = db_oauth
        .github_client_secret
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| config.github_client_secret.clone());

    if client_id.is_empty() || client_secret.is_empty() {
        return Err(ApiResponse::internal_error("GitHub 配置缺失".into()));
    }

    // 使用新的统一 OAuth 服务
    let oauth_service = OAuthService::new(
        Some(client_id),
        Some(client_secret),
        None, // QQ redirect_uri
    );

    let user_info = oauth_service
        .exchange_github_code(code)
        .await
        .map_err(|e| ApiResponse::internal_error(format!("GitHub API 调用失败: {e}")))?;

    Ok(convert_to_payload(user_info))
}

/// 处理 QQ 的 OAuth 交换逻辑
async fn handle_qq_logic(
    code: &str,
    config: &OAuthConfig,
    _db: &Database,
) -> Result<OAuthUserPayload, Json<ApiResponse<()>>> {
    // 使用新的统一 OAuth 服务
    let oauth_service = OAuthService::new(
        None, // GitHub credentials
        None,
        Some(config.qq_redirect_uri()),
    );

    let user_info = oauth_service
        .exchange_qq_code(code)
        .await
        .map_err(|e| ApiResponse::internal_error(format!("QQ API 调用失败: {e}")))?;

    Ok(convert_to_payload(user_info))
}

File: backend_ts/src/routes/auth/oauth.ts
import { Elysia, t } from "elysia";
import { getConfig } from "@/config";
import { db_read } from "@/lib/db";
import { responsePlugin } from "@/plugins/response";
import { convertToPayload, IdentityService } from "@/services/identity";
import { OAuthService } from "@/services/oauth";

const config = getConfig();
const DB_NAME = config.mongodb.database || "mx-space";

/**
 * OAuth 认证路由
 *
 * GET /oauth/:provider - 发起 OAuth 登录（重定向到 OAuth provider）
 * GET /oauth/:provider/callback - OAuth 回调处理
 *
 * 支持的 providers: github, qq
 */
export const oauthRoutes = new Elysia()
  .use(responsePlugin())

  // GET /oauth/:provider - 发起 OAuth 登录
  .get(
    "/oauth/:provider",
    async ({ params, error, set }) => {
      const { provider } = params;

      // 验证 provider
      const supportedProviders = ["github", "qq"];
      if (!supportedProviders.includes(provider.toLowerCase())) {
        set.status = 400;
        return error(400, `Unsupported OAuth provider: ${provider}`);
      }

      // TODO: 从数据库读取 OAuth 配置（优先级高于环境变量）
      // const dbOAuthConfig = await getOAuthConfigFromDB();

      let redirectUrl: string;

      switch (provider.toLowerCase()) {
        case "github": {
          const clientId = config.oauth?.github?.clientId || process.env.GITHUB_CLIENT_ID;
          const redirectUri = config.oauth?.github?.redirectUri || process.env.GITHUB_REDIRECT_URI;

          if (!clientId) {
            set.status = 500;
            return error(500, "GitHub OAuth not configured");
          }

          redirectUrl = `https://github.com/login/oauth/authorize?client_id=${clientId}&redirect_uri=${encodeURIComponent(redirectUri || "")}&scope=user:email`;
          break;
        }

        case "qq": {
          const redirectUri = config.oauth?.qq?.redirectUri || process.env.QQ_REDIRECT_URI;
          // QQ OAuth 使用统一的服务
          redirectUrl = `https://api-space.tnxg.top/oauth/qq/authorize?redirect=true&return_url=${encodeURIComponent(redirectUri || "")}`;
          break;
        }

        default:
          set.status = 400;
          return error(400, `Unsupported provider: ${provider}`);
      }

      // 重定向到 OAuth provider
      set.redirect = redirectUrl;
    },
  )

  // GET /oauth/:provider/callback - OAuth 回调
  .get(
    "/oauth/:provider/callback",
    async ({ params: { provider: _provider }, query, set, cookie }) => {
      const { code } = query;
      const provider = _provider.toLowerCase();

      // 验证参数
      if (!code) {
        // 重定向到前端并带上错误信息
        const frontendUrl = config.server?.frontendUrl || "http://localhost:3000";
        const errorMsg = encodeURIComponent("Missing authorization code");
        set.redirect = `${frontendUrl}/auth/callback?error=${errorMsg}`;
        return;
      }

      try {
        // 1. 获取数据库中的 OAuth 配置（优先级高于环境变量）
        const dbOAuthOptions = await db_read(DB_NAME, "options", { key: "oauth" }, {});
        const dbOAuth = dbOAuthOptions[0]?.value || {};

        // 2. 获取第三方用户信息
        let userInfo;

        if (provider === "github") {
          const clientId = dbOAuth.github_client_id || config.oauth?.github?.clientId || process.env.GITHUB_CLIENT_ID;
          const clientSecret = dbOAuth.github_client_secret || config.oauth?.github?.clientSecret || process.env.GITHUB_CLIENT_SECRET;

          if (!clientId || !clientSecret) {
            throw new Error("GitHub 配置缺失");
          }

          const oauthService = new OAuthService(clientId, clientSecret, undefined);
          userInfo = await oauthService.exchangeGitHubCode(code);
        } else if (provider === "qq") {
          const redirectUri = config.oauth?.qq?.redirectUri || process.env.QQ_REDIRECT_URI;
          const oauthService = new OAuthService(undefined, undefined, redirectUri);
          userInfo = await oauthService.exchangeQQCode(code);
        } else {
          throw new Error("不支持的提供商");
        }

        // 3. 转换为标准 Payload
        const payload = convertToPayload(userInfo);

        // 4. 使用 IdentityService 处理复杂的业务逻辑
        const identityService = new IdentityService();
        const [userId, isOwner, isNewUser] = await identityService.processOAuthLogin(payload);

        // 5. 颁发 JWT 令牌
        const token = await identityService.issueToken(userId, isOwner);

        // 6. 设置 HttpOnly Cookie（用于后端 API 鉴权）
        cookie.auth_token.set({
          value: token,
          httpOnly: true,
          secure: true, // 生产环境强制 HTTPS
          sameSite: "lax",
          path: "/",
          maxAge: 7 * 24 * 60 * 60, // 7 天
        });

        // 7. 重定向回前端页面
        const frontendUrl = config.server?.frontendUrl || "http://localhost:3000";
        const callbackUrl = `${frontendUrl}/auth/callback?token=${token}&new_user=${isNewUser}`;
        set.redirect = callbackUrl;
      } catch (error) {
        // 如果处理失败，重定向到前端并带上错误参数
        const frontendUrl = config.server?.frontendUrl || "http://localhost:3000";
        const errorMsg = error instanceof Error ? error.message : "OAuth 处理失败";
        set.redirect = `${frontendUrl}/auth/callback?error=${encodeURIComponent(errorMsg)}`;
      }
    },
    {
      query: t.Object({
        code: t.Optional(t.String()),
        state: t.Optional(t.String()),
      }),
    },
  );
